File: backero-backend/src/models/TaskApproval.js
const mongoose = require('mongoose');
const { APPROVAL_STATUS } = require('../utils/constants');

const taskApprovalSchema = new mongoose.Schema({
  organizationId: { type: mongoose.Schema.Types.ObjectId, ref: 'Organization', required: true, index: true },
  taskId: { type: mongoose.Schema.Types.ObjectId, ref: 'Task', required: true, index: true },
  requestedBy: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
  reviewedBy: { type: mongoose.Schema.Types.ObjectId, ref: 'User' },
  status: {
    type: String,
    enum: Object.values(APPROVAL_STATUS),
    default: APPROVAL_STATUS.PENDING,
    index: true,
  },
  requestNotes: { type: String },
  reviewNotes: { type: String },
  attachments: [{ url: String, name: String, type: String }],
  requestedAt: { type: Date, default: Date.now },
  reviewedAt: { type: Date },
  round: { type: Number, default: 1 },
  createdBy: { type: mongoose.Schema.Types.ObjectId, ref: 'User' },
  updatedBy: { type: mongoose.Schema.Types.ObjectId, ref: 'User' },
}, { timestamps: true });

taskApprovalSchema.index({ organizationId: 1, status: 1 });
taskApprovalSchema.index({ organizationId: 1, requestedBy: 1 });
taskApprovalSchema.index({ organizationId: 1, reviewedBy: 1 });

module.exports = mongoose.model('TaskApproval', taskApprovalSchema);

File: backero-backend-py/app/models/task_approval.py
"""
TaskApproval — ported from Task_Workflow's Mongoose `TaskApproval` model.
Drives the core request-completion -> approve/reject loop (see
approval.controller.js's approveTask/rejectTask, read directly during
planning). "Dept Hub" approval is a separate, unrelated concept (see
task.py's module docstring) and has no representation here.
"""

from __future__ import annotations

import uuid
from datetime import datetime
from enum import StrEnum

from sqlalchemy import DateTime, ForeignKey, Integer, Text
from sqlalchemy import Enum as SAEnum
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.core.database import Base
from app.models.mixins import TimestampMixin, UUIDPrimaryKeyMixin


class ApprovalStatus(StrEnum):
    PENDING = "pending"
    APPROVED = "approved"
    REJECTED = "rejected"
    CHANGES_REQUESTED = "changes_requested"


class TaskApproval(Base, UUIDPrimaryKeyMixin, TimestampMixin):
    __tablename__ = "task_approvals"

    organization_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("workflow_organizations.id"), nullable=False, index=True
    )
    task_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("tasks.id", ondelete="CASCADE"), nullable=False, index=True
    )
    requested_by_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("workflow_users.id"), nullable=False
    )
    reviewed_by_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("workflow_users.id"), nullable=True
    )
    status: Mapped[ApprovalStatus] = mapped_column(
        SAEnum(ApprovalStatus, name="approval_status", native_enum=True),
        nullable=False,
        default=ApprovalStatus.PENDING,
        index=True,
    )
    request_notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    review_notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    requested_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    reviewed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    round: Mapped[int] = mapped_column(Integer, nullable=False, default=1)
